exercises/book/A3-traits-generics/mod.md

```diff
@@ -0,0 +1,151 @@
+# Module A3 - Traits and generics
+
+
+## A3.2 Local Storage Vec
+In this exercise, we'll create a type called `LocalStorageVec`, which is generic list of items that resides either on the stack or the heap, depending on its size. If its size is small enough for items to be put on the stack, the `LocalStorageVec` buffer is backed by an array. `LocalStorageVec` is not only generic over the type  (`T`) of items in the list, but also by the size (`N`) of this stack-located array using a relatively new feature called 'const generics'. Once the `LocalStorageVec` contains more items than fit in the array, a heap based [`Vec`](https://doc.rust-lang.org/std/vec/struct.Vec.html) is allocated as space for the items to reside in.
+
+**Questions**
+1. When is such a data structure more efficient than a standard `Vec`?
+2. What are the downsides, compared to `Vec`?
+
+Open the `exercises/A3/2-local-storage-vec` crate. It contains a `src/lib.rs` file, meaning this crate is a library. `lib.rs` contains a number of tests, which can be run by calling `cargo test`. Don't worry if they don't pass or even compile right now: it's your job to fix that in this exercise. Most of the tests are commented out right now, to enable a step-by-step approach. **Before you begin, have a look at the code and the comments in there, they contain various helpful clues.**
+
+### A3.2.A Defining the type
+Currently, the `LocalStorageVec` `enum` is incomplete. Give it two variants: `Stack` and `Heap`. `Stack` contains two named fields, `buf` and `usize`. `buf` will be the array containing `N` items of type `T`. The `Heap` variant has an unnamed field containing a `Vec<T>`. If you've defined the `LocalStorageVec` variants correctly, running `cargo test` should output something like
+
+```txt
+running 1 test
+test test::it_compiles ... ignored, This test is just to validate the definition of `LocalStorageVec`. If it compiles, all is OK
+
+test result: ok. 0 passed; 0 failed; 1 ignored; 0 measured; 0 filtered out; finished in 0.00s
+```
+
+This test does (and should) not run, but is just there for checking your variant definition.
+
+
+### A4.2.B `impl`-ing `From<Vec<T>`
+
+Uncomment the test `it_from_vecs`, and add an implementation for `From<Vec<T>>` to `LocalStorageVec<T>`. To do so, copy the following code in your `lib.rs` file and replace the `todo!` macro invocation with your code that creates a heap-based `LocalStorageVec` containing the passed `Vec<T>`.
+
+```rust
+impl<T, const N: usize> From<Vec<T>> for LocalStorageVec<T, N> {
+    fn from(v: Vec<T>) -> Self {
+        todo!("Implement me");
+    }
+}
+```
+
+**Question** 
+1. How would you pronounce the first line of the code you just copied in English?*
```

Comment:
What  does this question try to make students reflect upon?

Reply:
It should be pretty trivial to answer the question, but I'd like for students to be able to reason about what the syntax means